### DISCUSSION TITLE
Dynamic file upload

### DISCUSSION THREAD

**Discussion by ghost** [abramenal/cypress-file-upload#267] [Q&A] [answered]
Currently, the only way to upload a file is from fixtures.
It would be great to upload files programmatically! 

In my case I use "xlsx" to create files like this:

 ```
const wb = XLSX.utils.book_new();
 const ws_name = "test";
 const ws_data = [...];
 const ws = XLSX.utils.aoa_to_sheet(ws_data);
   XLSX.utils.book_append_sheet(wb, ws, ws_name);
 XLSX.writeFile(wb, "example.xlsx");
```

As I want to test my code in terms of parsing, reading and validating different kinds of dynamic xlsx files, it would be great to somehow upload files accordingly without having to create tons of xlsx files into fixtures (taking a lot of disk space).

What do you think? Do have better ideas? Looking forward to hearing from you!!

**Comment by c32hedge** [comment, marked as answer]
I don't think this is quite what you were looking for, but I had a similar goal of not having to create tons of fixtures and found that if I could replicate the format of the file contents that `cy.fixture` yields, I didn't need a real file at all. Obviously, this is probably easier for my use case of simple `csv` and `txt` files, but I used `console.log` to see what value was being passed to my `cy.fixture` callback and came up with something like this:

``` javascript
cy.get('<my-file-input-selector>').attachFile({
  fileContent: 'Header1,Header2\nvalue A,value B\nvalue C,value D',
  fileName: 'LookMaNoRealFile.csv',
  mimeType: 'text/csv',
  encoding: 'utf-8'
});
```

**Comment by abramenal** [comment]
As mentioned, there is a way to attach raw file content – please see mentioned example, or related section of API docs ;)

**Reply by c32hedge** [reply]
@abramenal It's worth mentioning that the couple examples in the API docs still use `cy.fixture` to read the contents of a file, so it wasn't immediately obvious, at least to me, that I could do away with the real file entirely until I experimented with it.
